### PR TITLE
feat: implement 'dispatch namespaces'

### PR DESCRIPTION
**Pull request by aaronlisman**
Let me know if more, or more thorough, tests should be added.

The commands are hidden from help for now, as it's not widely available yet.

Besides the unit tests, I've tested manually on my own account and have been able to get the example below working.

=== commit message ===

This feature is for 'Workers for Platforms', and enables usage of a few
'namespaces' APIs to get the bare minimum example running.

New Commands:
  wrangler dispatch-namespace create <name>
  wrangler dispatch-namespace delete <name>

New Bindings:
  [[dispatch_namespaces]]
  name = "dispatcher"
  namespace = "my-namespace"

New Config:
  dispatch_namespace = "my-namespace"

Example usage would be to create a namespace with `dispatch-namespace
create my-namespace`, upload a worker *to* that namespace by including
`dispatch_namespace = "my-namespace"` in a worker's toml, and then
upload a worker with a 'namespace' binding that can be used to dispatch
to the worker we just uploaded (by including a `[[dispatch_namespaces]]`
binding in its `wrangler.toml`).

After doing those three things, you should be able to route to the last
worker and call out to the first worker using the new binding.

### CODE REVIEW

File: packages/wrangler/src/dispatch/index.ts [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    deleteNamespace,
+    deleteNamespaceCommand,
+
+    // still to be implemented in API:
+    // fetchNamespace,
+    // fetchNamespaceCommand,
+    //
+    // listNamespaces,
+    // listNamespacesCommand,
+}
```
Comment (JacobMGEvans) [inline packages/wrangler/src/dispatch/index.ts:119]:
```suggestion
}

```

File: packages/wrangler/src/dispatch/delete-namespace.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+export default async function deleteNamespace(props: Props): Promise<void> {
+    await fetchResult(
+        namespacePath(props.accountId, props.namespaceName),
+        {
+            method: "DELETE",
+        }
+    )
+
+    logger.log(`Deleted namespace ${props.namespaceName}.`)
+}
```
Comment (JacobMGEvans) [inline packages/wrangler/src/dispatch/delete-namespace.ts:21]:
```suggestion
}

```

File: packages/wrangler/src/dispatch/create-namespace.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+            method: "POST",
+            headers: {
+                "Content-Type": "application/json",
+            },
+            body: JSON.stringify(body),
+        },
+    )
+
+    logger.log(`Created namespace ${props.namespaceName} with ID ${namespace.namespace_id}`)
+}
```
Comment (JacobMGEvans) [inline packages/wrangler/src/dispatch/create-namespace.ts:31]:
```suggestion
}

```

File: packages/wrangler/src/dispatch/index.ts [hunk 4]
```diff
@@ -0,0 +1,119 @@
+import { readConfig } from "../config";
```
Comment (JacobMGEvans) [inline packages/wrangler/src/dispatch/index.ts:1]: `/dispatch/index.ts` I think should be `/dispatch/dispatch.ts` this is a personal preference.

File: packages/wrangler/src/publish.ts [hunk 5]
```diff
@@ -251,8 +255,25 @@ export default async function publish(props: Props): Promise<void> {
...
+      (routes && routes.length > 0)
+      || (triggers && triggers.length > 0)
+    )
+  ) {
+    logger.warn(
+      `Publishing to namespace '${namespaceName}': one or more routes or cron triggers are defined that will not be deployed.`
+    );
+  }
+
   // deployToWorkersDev defaults to true only if there aren't any routes defined
```
Comment (JacobMGEvans) [inline packages/wrangler/src/publish.ts:270]:
```suggestion
  // deployToWorkersDev defaults to true only if there aren't any routes or namespaces defined
```

File: packages/wrangler/src/publish.ts [hunk 6]
```diff
@@ -290,14 +311,22 @@ export default async function publish(props: Props): Promise<void> {
   }
 
   const destination = props.outDir ?? (await tmp.dir({ unsafeCleanup: true }));
-  const envName = props.env ?? "production";
+  const envName = props.env ?? DEFAULT_ENV_NAME;
+
+  if (namespaceName && envName != DEFAULT_ENV_NAME) {
```
Comment (JacobMGEvans) [inline packages/wrangler/src/publish.ts:316]:
```suggestion
  if (namespaceName && envName !== DEFAULT_ENV_NAME) {
```

Comment (JacobMGEvans) [reply packages/wrangler/src/publish.ts:316]: `!==` is closer to what you are familiar with `!=` means in Go

File: packages/wrangler/src/publish.ts [hunk 7]
```diff
@@ -290,14 +311,22 @@ export default async function publish(props: Props): Promise<void> {
...
   const notProd = Boolean(!props.legacyEnv && props.env);
   const workerName = notProd ? `${scriptName} (${envName})` : scriptName;
-  const workerUrl = notProd
-    ? `/accounts/${accountId}/workers/services/${scriptName}/environments/${envName}`
-    : `/accounts/${accountId}/workers/scripts/${scriptName}`;
+  const workerUrl = namespaceName
+    ? `/accounts/${accountId}/workers/dispatch/namespaces/${namespaceName}/scripts/${scriptName}`
+    : notProd
+      ? `/accounts/${accountId}/workers/services/${scriptName}/environments/${envName}`
+      : `/accounts/${accountId}/workers/scripts/${scriptName}`;
```
Comment (JacobMGEvans) [inline packages/wrangler/src/publish.ts:329]: Maybe this shouldn't be a ternary anymore 😅

Comment (JacobMGEvans) [reply packages/wrangler/src/publish.ts:329]: Maybe a nested `if/else` 
```suggestion
if (const workerUrl = namespaceName) {
	`/accounts/${accountId}/workers/dispatch/namespaces/${namespaceName}/scripts/${scriptName}`
} else {
	if (notProd) {
		`/accounts/${accountId}/workers/services/${scriptName}/environments/${envName}`
	} else {
		`/accounts/${accountId}/workers/scripts/${scriptName}`;
	}
}
```

Comment (threepointone) [reply packages/wrangler/src/publish.ts:329]: This code doesn't work, if/else doesn't "return" statements. The ternary is fine.

Comment (JacobMGEvans) [reply packages/wrangler/src/publish.ts:329]: I should have had the returns lol sounds good with the ternary just curious what our cutoff on chained ternaries are then?

Comment (aaronlisman) [reply packages/wrangler/src/publish.ts:329]: It did seem a bit... 'verbose' to me. I'll try breaking it out into a function.
This is used elsewhere too, I think? I guess the other place I saw it was tests though.

File: packages/wrangler/src/dispatch/index.ts (hunk 1 above)
Comment (JacobMGEvans) [reply packages/wrangler/src/dispatch/index.ts]: I agree!

**Comment by threepointone** [comment]
I just had a conversation with @aaronlisman about this PR (and Workers for Platforms in general), and we're going to park this PR for a bit. It served its purpose of demonstrating the approach for some people. We're also going to explore alternate integration strategies; (we should probably build libraries as well). We'll revisit this after some brainstorming. Really exciting!

**Review by threepointone — CHANGES_REQUESTED** [review]
Holding back from landing as discussed.

**Comment by threepointone** [comment]
Closing this so it doesn't show in our active PRs list, but happy to revisit and reopen if needed.

**Comment by threepointone** [comment]
We should probably have a PR with just the namespace creation, that will still be useful for wrangler to directly provide. Additionally, it should also have namespace deletion (and maybe also renaming?)

**Comment by threepointone** [comment]
The bindings also make sense. I guess the only one that needs a non-wrangler story is publishing to a namespace.

**Pull request closed without merging**